pkg/store/limiter.go

```diff
@@ -82,16 +86,94 @@ func NewChunksLimiterFactory(limit uint64) ChunksLimiterFactory {
+	SamplesPerRequest uint64
+}
+
+func (l *SeriesSelectLimits) RegisterFlags(cmd extkingpin.FlagClause) {
+	cmd.Flag("store.grpc.series-limit", "The maximum series allowed for a single Series request. The Series call fails if this limit is exceeded. 0 means no limit.").Default("0").Uint64Var(&l.SeriesPerRequest)
+	cmd.Flag("store.grpc.samples-limit", "The maximum samples allowed for a single Series request, The Series call fails if this limit is exceeded. 0 means no limit. NOTE: For efficiency the limit is internally implemented as 'chunks limit' considering each chunk contains a maximum of 120 samples.").Default("0").Uint64Var(&l.SamplesPerRequest)
+}
+
+var _ storepb.StoreServer = &limitedStoreServer{}
+
+// limitedStoreServer is a storepb.StoreServer that can apply series and sample limits against individual Series requests.
+type limitedStoreServer struct {
+	storepb.StoreServer
+	newSeriesLimiter      SeriesLimiterFactory
+	newSamplesLimiter     ChunksLimiterFactory
+	failedRequestsCounter *prometheus.CounterVec
+}
+
+// NewLimitedStoreServer creates a new limitedStoreServer.
+func NewLimitedStoreServer(store storepb.StoreServer, reg prometheus.Registerer, selectLimits SeriesSelectLimits) storepb.StoreServer {
+	return &limitedStoreServer{
+		StoreServer:       store,
+		newSeriesLimiter:  NewSeriesLimiterFactory(selectLimits.SeriesPerRequest),
+		newSamplesLimiter: NewChunksLimiterFactory(selectLimits.SamplesPerRequest),
+		failedRequestsCounter: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "thanos_store_selects_dropped_total",
+			Help: "Number of select queries that were dropped due to configured limits.",
+		}, []string{"reason"}),
+	}
+}
+
+func (s *limitedStoreServer) Series(req *storepb.SeriesRequest, srv storepb.Store_SeriesServer) error {
+	seriesLimiter := s.newSeriesLimiter(s.failedRequestsCounter.WithLabelValues("series"))
+	chunksLimiter := s.newSamplesLimiter(s.failedRequestsCounter.WithLabelValues("chunks"))
+	limitedSrv := newLimitedServer(srv, seriesLimiter, chunksLimiter)
+	if err := s.StoreServer.Series(req, limitedSrv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ storepb.Store_SeriesServer = &limitedServer{}
+
+// limitedServer is a storepb.Store_SeriesServer that tracks statistics about sent series.
+type limitedServer struct {
+	storepb.Store_SeriesServer
+	seriesLimiter  SeriesLimiter
+	samplesLimiter ChunksLimiter
+}
+
+func newLimitedServer(upstream storepb.Store_SeriesServer, seriesLimiter SeriesLimiter, chunksLimiter ChunksLimiter) *limitedServer {
+	return &limitedServer{
+		Store_SeriesServer: upstream,
+		seriesLimiter:      seriesLimiter,
+		samplesLimiter:     chunksLimiter,
+	}
+}
+
+func (i *limitedServer) Send(response *storepb.SeriesResponse) error {
+	series := response.GetSeries()
+	if series == nil {
+		return i.Store_SeriesServer.Send(response)
+	}
+
+	if err := i.seriesLimiter.Reserve(1); err != nil {
+		return errors.Wrapf(err, "failed to send series")
```

Comment:
I wonder once this error bubbles up, if we should use a specific gRPC code. I think currently we use `Unknown` for all errors, but perhaps in this case we could use `ResourceExhausted` or something similar. Could be useful when debugging issues to see this directly in response or server metrics / span errors.

But can be done additionally, not a blocker.

Reply:
I think it's a good idea, but I'm concerned it might grow the scope of the PR even further. Right now we have no way to distinguish error types on the client side, and we mostly rely on surfacing the actual message of the user. In this case they would see something like `failed to send series: limit X violated (got Y)`